zenoh-flow/src/traits.rs

```diff
@@ -22,49 +22,194 @@ use std::any::Any;
+/// not be able to handle the state.
+///
+/// It can be easily derived.
+///
+/// Example:
+///
+/// ```no_run
+/// use zenoh_flow::zenoh_flow_derive::ZFState;
+/// #[derive(Debug, Clone, ZFState)]
+/// pub struct MyState;
+/// ```
+///
+
 pub trait ZFState: Debug + Send + Sync {
+    /// Donwcast as reference to [`Any`](`Any`)
     fn as_any(&self) -> &dyn Any;
+
+    /// Donwcast as mutable reference to [`Any`](`Any`)
     fn as_mut_any(&mut self) -> &mut dyn Any;
 }
 
+/// The `Node` trait represents a generic node in the data flow graph.
+/// It contains functions that are common between Operator, Sink and Source.
+/// It has to be implemented for each node within a graph.
 pub trait Node {
+    /// This method is used to initialize the state of the node.
+    /// It is called by the Zenoh Flow runtime when initializing the data flow
+    /// graph.
+    /// An example of node state is files that should be opened, connection
+    /// to devices or internal configuration.
+    ///
+    /// # Errors
+    /// If it fails to initialize an error variant will be returned.
     fn initialize(&self, configuration: &Option<Configuration>) -> ZFResult<State>;
 
+    /// This method is used to finalize the state of the node.
+    /// It is called by the Zenoh Flow runtime when tearing down the data
+    /// flow graph.
+    ///
+    /// An example of node state to finalize is files to be closed,
+    /// clean-up of libraries, or devices.
+    ///
+    /// # Errors
+    /// If it fails to finalize an error variant will be returned.
     fn finalize(&self, state: &mut State) -> ZFResult<()>;
 }
 
+/// The `Operator` trait represent an Operator inside Zenoh Flow.
 pub trait Operator: Node + Send + Sync {
+    /// This method is called when data is received on one or more inputs.
+    /// The result of this method is use as discriminant to trigger the
+    /// operator's run function.
+    /// The operator can access to its context and its state during execution.
+    ///
+    /// The received data is provided as [`InputToken`](`InputToken`) that
+    /// represent the state of the associated port.
+    /// Based on the tokens and on the data users can decide if trigger
+    /// the run or not.
+    /// The commodity function [`default_input_rule`](`default_input_rule`)
+    /// can be used if the operator should be triggered based on KPN rules.
+    ///
+    /// # Errors
+    /// If something goes wrong during execution an error
+    /// variant will be returned.
     fn input_rule(
         &self,
         context: &mut Context,
         state: &mut State,
         tokens: &mut HashMap<PortId, InputToken>,
     ) -> ZFResult<bool>;
 
+    /// This method is the actual one processing the data.
+    /// It is triggered based on the result of the `input_rule`.
+    /// As operators are computing over data,
+    /// *I/O should not be done in the run*.
+    ///
+    /// The operator can access to its context and its state during execution.
```

Comment:
An operator can access its context and state during the execution of this function.